### PR TITLE
ARROW-2480: [C++] Enable casting the value of a decimal to int32_t or int64_t

### PR DESCRIPTION
**Pull request by cpcloud**

### CODE REVIEW

File: cpp/src/arrow/util/decimal.h
Comment (pitrou) [inline cpp/src/arrow/util/decimal.h]: IMHO it would be clearer, more explicit, and more consistent with the other methods if this were called `ToInteger`.

**Comment by pitrou** [comment]
@cpcloud  Do you plan to fix the lint problems here?

**Comment by cpcloud** [comment]
Yes

**Comment by xhochy** [comment]
@pcmoritz What's the state of this? Can you rebase and fix the compilation errors?

**Comment by wesm** [comment]
Other Phil @cpcloud :)

File: cpp/src/arrow/util/decimal.h
Comment (pitrou) [inline cpp/src/arrow/util/decimal.h]: Seems to me the `EnableIfIsOneOf` restriction isn't useful and removing it would actually fix the compile errors.

**Comment by pitrou** [comment]
Rebased.

**Comment by pitrou** [comment]
Ok, it _seems_ the weird compile error on gcc 4.8 was due to a missing header include for `EnableIfIsOneOf`! Why it worked on other compilers I don't know...

**Review by pitrou — APPROVED** [review]
LGTM.

**Comment by pitrou** [comment]
Rebasing after PR #2326 is merged will hopefully fix the failures.

**Review by xhochy — APPROVED** [review]
+1, LGTM

**Comment by pitrou** [comment]
What is the status on merges currently? Should we wait for the release?

**Comment by xhochy** [comment]
I think we should wait until the release is done, @wesm ?

**Comment by wesm** [comment]
I've continued to merge commits as with the last several releases. I think we have too many PRs to put merging on hold for 3-4 days. We can rebase master once the release vote closes

**Comment by wesm** [comment]
At some point we may have to start making releases out of branches instead of having the release commits in mainline

**Pull request closed without merging**
